src/outputs.rs

```diff
@@ -0,0 +1,168 @@
+    }
+}
+
+#[derive(Debug)]
+pub struct Service {
+    // Associates a connector ID (e.g. HDMI-1) to a state
+    output_channel: BroadcastChannel<MonitorEvent>,
+}
+
+impl Service {
+    pub(crate) fn new(ironbar: &Rc<Ironbar>) -> Self {
+        let output_channel = broadcast::channel(32);
+
+        let mut rx_wl_outputs = ironbar.clients.borrow_mut().wayland().subscribe_outputs();
+
+        let monitors = Arc::new(Mutex::new(HashMap::new()));
+
+        // listen to wayland events
+        {
+            let monitors = monitors.clone();
+            let output_tx = output_channel.0.clone();
+
+            glib::spawn_future_local(async move {
+                while let Ok(event) = rx_wl_outputs.recv().await {
+                    debug!("Wayland output event: {event:?}");
+                    if let Some(name) = &event.output.name {
+                        let mut guard = monitors.lock().unwrap();
+                        let entry = guard.entry(name.clone()).or_insert_with(|| MonitorEvent {
+                            connector: name.clone(),
+                            state: MonitorState::Disconnected,
+                            tx: output_tx.clone(),
+                        });
+                        match event.event_type {
+                            wayland::OutputEventType::New => entry.connect_wayland(&event.output),
+                            wayland::OutputEventType::Destroyed => entry.disconnect(),
+                            wayland::OutputEventType::Update => {}
+                        };
+                    }
+                }
+            });
+        }
+
+        // listen to GDK events
+        {
+            let output_tx = output_channel.0.clone();
+            let monitors = monitors.clone();
+
+            let display = get_display();
+            display
+                .monitors()
+                .connect_items_changed(move |list, position, removed, added| {
+                    debug!("GDK event: +{added}, -{removed}");
+                    for added_idx in position..position + added {
+                        if let Some(monitor) = list.item(added_idx).and_downcast::<Monitor>() {
+                            let output_tx = output_tx.clone();
+                            let monitors = monitors.clone();
+
+                            /*
+                             * At this point, we have a `gdk::Monitor`. However, the object initially
+                             * has all its fields `None`, including `connector`.
+                             *
+                             * We have to listen for the notify event for the `connector` being set.
+                             */
+                            monitor.connect_notify(Some("connector"), move |m, _| {
+                                if let Some(connector) = m.connector() {
+                                    monitors
+                                        .lock()
+                                        .unwrap()
```

Comment:
`unwrap` should be avoided in production code. You can instead use the `lock!()` macro (`crate::lock`):


```suggestion
                                    lock!(monitors)
```

Under the hood this will still panic if the lock fails (this should never happen), but it uses `expect` which presents a clearer message.